### PR TITLE
[pigeon] Added a contributor's guide.

### PR DESCRIPTION
**Pull request by gaaclarke**
* Added fix for CI by tagging what version of style checks we use.
* Updated the formatting of the dart generator to please a change in CI

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated `CHANGELOG.md` to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the `#hackers-new` channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: packages/pigeon/CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# Pigeon Contributor's Guide
+
+**author:** gaaclarke\
+**updated:** 4/6/21
```
Comment (stuartmorgan) [inline packages/pigeon/CONTRIBUTING.md:4]: Why is there an author and update date, given that both are going to be accurately captured in the git history? This seems likely to become stale.

File: packages/pigeon/CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+## Description
+
+Pigeon is a code generation tool that adds type safety to Flutter’s Platform
+Channels.  This document serves as an overview of how it functions to help
+people who would like to contribute to the project.
```
Comment (stuartmorgan) [inline packages/pigeon/CONTRIBUTING.md:10]: This seems redundant with the README?

File: packages/pigeon/CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+## State Diagram
+
+Pigeon generates a temporary file in it's LaunchIsolate, then launches another
```
Comment (stuartmorgan) [inline packages/pigeon/CONTRIBUTING.md:14]: Nit: its

File: packages/pigeon/CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  [dart_generator_test.dart](./test/dart_generator_test.dart)
+* Compilation tests -  These tests generate code, then attempt to compile that
+  code.  These are tests are much slower than unit tests, but not as slow as
+  integration tests.  These tests are typically run against the Pigeon files in
+  [pigeons](./pigeons).
+* Integration tests - These tests generate code, then compile the generated
+  code, then execute the generated code.  It can be thought of as unit-tests run
+  against the generated code.  Examples: [platform_tests](./platform_tests)
+
+## Eminent Plans
```
Comment (stuartmorgan) [inline packages/pigeon/CONTRIBUTING.md:68]: Do you mean "imminent"?

Also, I'm not sure what exactly this section is telling me if I'm someone interested in contributing. Are these things you're looking for volunteers to implement? Things that block other works so potential contributors should wait for? Something else?

File: packages/pigeon/run_tests.sh [hunk 5]
```diff
@@ -129,7 +130,7 @@ if [ ! -f "ci/$JAVA_FORMATTER" ]; then
   curl -L https://github.com/google/google-java-format/releases/download/google-java-format-1.3/$JAVA_FORMATTER > "ci/$JAVA_FORMATTER"
 fi
 if [ ! -f "ci/$GOOGLE_CHECKS" ]; then
-  curl -L https://raw.githubusercontent.com/checkstyle/checkstyle/master/src/main/resources/$GOOGLE_CHECKS > "ci/$GOOGLE_CHECKS"
+  curl -L https://raw.githubusercontent.com/checkstyle/checkstyle/$GOOGLE_CHECKS_VERSION/src/main/resources/$GOOGLE_CHECKS > "ci/$GOOGLE_CHECKS"
```
Comment (stuartmorgan) [inline packages/pigeon/run_tests.sh:133]: How is this related to the contributing guide? I'm not sure what this is based on the PR description.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/pigeon/CONTRIBUTING.md (hunk 1 above)
Comment (gaaclarke) [reply packages/pigeon/CONTRIBUTING.md]: Yea, fair enough, removed.

File: packages/pigeon/CONTRIBUTING.md (hunk 2 above)
Comment (gaaclarke) [reply packages/pigeon/CONTRIBUTING.md:7]: Yea, it's one sentence of redundancy that builds up context for the one sentence description.  I think it helps without being too burdensome to those that already have the context.

File: packages/pigeon/CONTRIBUTING.md (hunk 3 above)
Comment (gaaclarke) [reply packages/pigeon/CONTRIBUTING.md]: done

File: packages/pigeon/run_tests.sh (hunk 5 above)
Comment (gaaclarke) [reply packages/pigeon/run_tests.sh:133]: updated description, it's only related in the fact that this is the first PR that has run into the problem of drift in the CI script

File: packages/pigeon/CONTRIBUTING.md (hunk 4 above)
Comment (gaaclarke) [reply packages/pigeon/CONTRIBUTING.md]: The idea is just to give people a heads up on big changes that could affect contributions.  The first bullet point could be an important signal to people that want to contribute in a few ways:   1) Don't invest a lot of effort in the the part of the code that will be thrown away  2) If you want to implement a certain feature, it might be easier after that migration.  The second point gives people heads up that if they are contributing to Android the review process might be a bit heavier and identifies a need for contribution, sure.

File: packages/pigeon/CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+## Description
+
+Pigeon is a code generation tool that adds type safety to Flutter’s Platform
+Channels.  This document serves as an overview of how it functions to help
+people who would like to contribute to the project.
+
+## State Diagram
+
+Pigeon generates a temporary file in its LaunchIsolate, then launches another
```
Comment (xster) [inline packages/pigeon/CONTRIBUTING.md:11]: You might have to define these words for contributors.

File: packages/pigeon/CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+isolate that uses dart:mirrors to parse the generated file, creating an AST,
+then running code generators with that AST.
+
+![State Diagram](./docs/pigeon_state.png)
+
+## Generated Source Code Example
+
+This is what the generated code looks like:
```
Comment (xster) [inline packages/pigeon/CONTRIBUTING.md:19]: Not sure I understood this section

File: packages/pigeon/CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  code.  These are tests are much slower than unit tests, but not as slow as
+  integration tests.  These tests are typically run against the Pigeon files in
+  [pigeons](./pigeons).
+* Integration tests - These tests generate code, then compile the generated
+  code, then execute the generated code.  It can be thought of as unit-tests run
+  against the generated code.  Examples: [platform_tests](./platform_tests)
+
+## Imminent Plans
+
+* Migrate to Dart Analyzer for AST generation ([issue
```
Comment (xster) [inline packages/pigeon/CONTRIBUTING.md:67]: This seems like a high maintenance text. Maybe just link to your Pigeon next steps doc or to a GitHub project under the flutter/flutter or flutter/packages?

File: packages/pigeon/CONTRIBUTING.md (hunk 6 above)
Comment (gaaclarke) [reply packages/pigeon/CONTRIBUTING.md:11]: I defined LaunchIsolate and PigeonIsolate.

File: packages/pigeon/CONTRIBUTING.md (hunk 7 above)
Comment (gaaclarke) [reply packages/pigeon/CONTRIBUTING.md]: I expanded on the description sentence and moved it down lower in the document since it's not as important.

File: packages/pigeon/CONTRIBUTING.md (hunk 8 above)
Comment (gaaclarke) [reply packages/pigeon/CONTRIBUTING.md:68]: I agree it isn't ideal but this is where I have people's attention that want to contribute and I think the 2 points are important (see my response to Stuart above).  Hopefully 1 year from now the need for this section would be eliminated since we won't have any big changes coming and the testing will be even more mature.

File: packages/pigeon/CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+Pigeon is a code generation tool that adds type safety to Flutter’s Platform
+Channels.  This document serves as an overview of how it functions to help
+people who would like to contribute to the project.
+
+## State Diagram
+
+Pigeon generates a temporary file in its _LaunchIsolate_, the isolate that is
+spawned to run `main()`, then launches another isolate, _PigeonIsolate_, that
+uses `dart:mirrors` to parse the generated file, creating an AST, then running
```
Comment (xster) [inline packages/pigeon/CONTRIBUTING.md:13]: define ast too since it might not be familiar to contributors

File: packages/pigeon/CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+  code.  These are tests are much slower than unit tests, but not as slow as
+  integration tests.  These tests are typically run against the Pigeon files in
+  [pigeons](./pigeons).
+* Integration tests - These tests generate code, then compile the generated
+  code, then execute the generated code.  It can be thought of as unit-tests run
+  against the generated code.  Examples: [platform_tests](./platform_tests)
+
+## Generated Source Code Example
+
+This is what the generated code that the _PigeonIsolate_ executes looks like:
```
Comment (xster) [inline packages/pigeon/CONTRIBUTING.md:51]: I still don't understand what this section means :)

**Review by xster — APPROVED** [review]

File: packages/pigeon/CONTRIBUTING.md (hunk 9 above)
Comment (gaaclarke) [reply packages/pigeon/CONTRIBUTING.md]: Added link.

File: packages/pigeon/CONTRIBUTING.md (hunk 10 above)
Comment (gaaclarke) [reply packages/pigeon/CONTRIBUTING.md]: Ok, reworked it again.  It won't make sense unless you've read the State Diagram section, linked to it.
